accepted/00063-hub-iss.md

```diff
@@ -0,0 +1,246 @@
+  - transfer proper could be done via `rsync`, which provides transparent compression and delta encoding
+- all new code will be typed Python 3, connecting to Postgres via psycopg2 and no dependencies to the old Python stack
+- inspection of the schema can be implemented via `SELECT` queries on Postgres's internal tables
+  - current and new implementations would thus be completely independent, so could coexist for an initial period of time, as the new implementation matures
+
+## Impact on existing components and users
+All code would be new in a new component, so no change to existing components is expected. No impact on existing users is expected at all until the old implementation is dropped.
+
+When that happens, there is a chance of regressions for ISS users, which represent a small minority of total users.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+Syncing content between different versions of Uyuni on the ISS Master and on the ISS Slave may not work, if the database schema changes in non-compatible ways.
+
+This risk is fundamentally unavoidable at a conceptual level, although different mitigations can be put in place and the current implementation actually allows for a broader set of mitigations than the proposal in this RFC can allow.
+
+In particular:
+
+1. Some changes in the database structure could be "internal only", not really changing the nature of the information that's being exchanged. In those cases, the current implementation could make use of the fact the the XML representation does not change, thus the ISS Slave continues to operate normally
+2. Other changes would be reflected in the XML structure, but still not create problems because, for example, the current implementation ignores unknown tags. This is the case for completely new elements that the ISS Slave would just not import (which is most likely conceptually correct, as tables would also be missing)
+3. Finally, yet other changes could really introduce a backwards compatibility problem and, since the protocol is versioned, [the import program will send an alert and abort](https://github.com/SUSE/spacewalk/blob/Manager-3.2/backend/satellite_tools/xmlSource.py#L271-L298).
+
+With the proposal in this RFC, it is possible to implement protection from case 3. via a similar versioning mechanism, partially against 2. ignoring missing tables but only offer limited resilience against case 1., as we would not have a real database-agnostic representation of data.
+
+pl/SQL could still be used, as a last resort, to apply different commands to different ISS Slave versions, but that would be hardly maintainable, so it would only be suggested to handle critical situations temporarily.
+
+General recommendation would be to upgrade an ISS Master, then upgrade ISS Slaves before attempting any new Inter-Server Synchronization. Another RFC will detail how Hub could make such upgrades easier.
+
```

Comment:
With SQL export/import we limit us to stay on postgresql . Another DB might not be possible because of different SQL syntax and features.

We do not plan to support something different. So this is not a big problem.

Reply:
I should have made this evident in the Drawbacks section, added.